src/Microsoft.ML.Transforms/CategoricalHashTransform.cs

```diff
@@ -135,100 +136,123 @@ public sealed class Arguments : TransformInputBase
+
+            var columns = new List<CategoricalHashEstimator.ColumnInfo>();
+            foreach (var column in args.Column)
             {
-                h.CheckValue(args, nameof(args));
-                h.CheckValue(input, nameof(input));
-                h.CheckUserArg(Utils.Size(args.Column) > 0, nameof(args.Column), "Columns must be specified");
-                if (args.HashBits < 1 || args.HashBits >= NumBitsLim)
-                    throw h.ExceptUserArg(nameof(args.HashBits), "Number of bits must be between 1 and {0}", NumBitsLim - 1);
-
-                // creating the Hash function
-                var hashArgs = new HashTransformer.Arguments
-                {
-                    HashBits = args.HashBits,
-                    Seed = args.Seed,
-                    Ordered = args.Ordered,
-                    InvertHash = args.InvertHash,
-                    Column = new HashTransformer.Column[args.Column.Length]
-                };
-                for (int i = 0; i < args.Column.Length; i++)
-                {
-                    var column = args.Column[i];
-                    if (!column.TrySanitize())
-                        throw h.ExceptUserArg(nameof(Column.Name));
-                    h.Assert(!string.IsNullOrWhiteSpace(column.Name));
-                    h.Assert(!string.IsNullOrWhiteSpace(column.Source));
-                    hashArgs.Column[i] = new HashTransformer.Column
-                    {
-                        HashBits = column.HashBits,
-                        Seed = column.Seed,
-                        Ordered = column.Ordered,
-                        Name = column.Name,
-                        Source = column.Source,
-                        InvertHash = column.InvertHash,
-                    };
-                }
+                var col = new CategoricalHashEstimator.ColumnInfo(
+                    column.Source ?? column.Name,
+                    column.Name,
+                    column.OutputKind ?? args.OutputKind,
+                    column.HashBits ?? args.HashBits,
+                    column.Seed ?? args.Seed,
+                    column.Ordered ?? args.Ordered,
+                    column.InvertHash ?? args.InvertHash);
+                columns.Add(col);
+            }
+            return new CategoricalHashEstimator(env, columns.ToArray()).Fit(input).Transform(input) as IDataTransform;
+        }
+
+        private readonly TransformerChain<ITransformer> _transformer;
 
-                return CreateTransformCore(
-                    args.OutputKind, args.Column,
-                    args.Column.Select(col => col.OutputKind).ToList(),
-                    HashTransformer.Create(h, hashArgs, input),
-                    h,
-                    args);
+        public CategoricalHashTransform(HashEstimator hash, IEstimator<ITransformer> keyToVector, IDataView input)
+        {
+            var chain = hash.Append(keyToVector);
+            _transformer = chain.Fit(input);
```

Comment:
this code belongs to the caller I think. I don't feel comfortable having a public ctor for transformer that takes estimators as args. Also, there should not be a public ctor for the transformer

Reply:
Ok, good enough.

---
In reply to: [221013014](https://github.com/dotnet/machinelearning/pull/1033#discussion_r221013014) [](ancestors = 221013014,220399096)